blocks/randompruner.go

```diff
@@ -206,18 +260,19 @@ func (pruner *RandomPruner) prune(ctx context.Context, bytesToPrune uint64) erro
 	const cidPadLength = 64
 	writer := bufio.NewWriter(tmpFile)
 	cidCount := 0
-	for cid := range allCids {
-		if ctx.Err() != nil {
-			return ctx.Err()
-		}
 
-		// Don't consider pinned blocks for deletion
-		if pruner.isPinned(cid) {
+	pruner.allCidsLk.Lock()
+	for cid, status := range pruner.allCids {
+
+		if status.pinned && time.Since(status.pinTime) < pruner.pinDuration {
 			continue
 		}
 
+		status.pinned = false
+
 		paddedCidStr := fmt.Sprintf("%-*s", cidPadLength, cid.String())
```

Comment:
I hadn't noticed this before but we're assuming all CIDs are <= 64 characters for this to work.. that's a pretty brave assumption, also a nice attack if someone cared to try. As soon as we have 1 CID in the list that is longer than it's going to break the algorithm because everything below it will have the wrong offset.

This isn't a "fix", but it at least stops breakage:

```suggestion
		cs := cid.String()
		if len(cs) > cidPadLength {
			cs = cs[:cidPadLength]
		}
		paddedCidStr := fmt.Sprintf("%-*s", cidPadLength, cs)
```

still open to abuse because it means long CIDs don't get deleted at all, but better than screwing up the entire deletion process.

Unless there's a CID length limitation with bitswap I'm not aware of?